File: python/arrays/0015-3sum.py
from typing import List

def threeSum(nums: List[int]) -> List[List[int]]:
    response = []
    nums.sort()
    
    for i, n in enumerate(nums):
        if i > 0 and n == nums[i - 1]:
            continue
            
        l, r = i + 1, len(nums) - 1
        while l < r:
            s = n + nums[l] + nums[r]
            if s >  0:
                r -= 1
            elif s < 0:
                l += 1
            else:
                response.append([n, nums[l], nums[r]])
                
                l += 1
                while nums[l] == nums[l - 1] and l < r:
                    l += 1
    
    return response

File: rust/arrays/0015-3sum.rs
fn three_sum(nums: &mut [i32]) -> Vec<Vec<i32>> {
    let mut ans = vec![];

    nums.sort();
    for (i, &num) in nums.iter().enumerate() {
        if i > 0 && num == nums[i-1] { continue };

        let target = -num;
        if target < 0 { break; }

        let (mut left, mut right) = (i+1, nums.len()-1);
        while left < right {
            let sum = nums[left] + nums[right];
            if sum > target {
                right -= 1;
            } else if sum < target {
                left += 1;
            } else {
                ans.push(vec![num, nums[left], nums[right]]);
                left += 1;

                while nums[left] == nums[left-1] && left < right {
                    left += 1;
                }
            }
        }
    }

    ans
}